projects/ad2s1210_iio/app/ad2s1210_iio.c

```diff
@@ -0,0 +1,619 @@
+
+	case FREQ_AVAIL_ATTR_ID:
+		return snprintf(buf, len, "[%d %d %d]", AD2S1210_MIN_EXCIT,
+				250, AD2S1210_MAX_EXCIT);
+	default:
+
+		break;
+	}
+
+	return len;
+}
+
+static int iio_ad2s1210_attr_set(void *device, char *buf, uint32_t len,
+				 const struct iio_ch_info *channel,
+				 intptr_t priv)
+{
+	struct ad2s1210_dev *dev = (struct ad2s1210_dev *) device;
+	uint16_t fexcit;
+	uint8_t hysteresis;
+	int ret;
+
+	switch (priv) {
+	case RAW_ATTR_ID:
+	case SCALE_ATTR_ID:
+	case LABEL_ATTR_ID:
+	case HYSTERESIS_AVAILABLE_ATTR_ID:
+	case FREQ_AVAIL_ATTR_ID:
+		/* All read-only attributes */
+		break;
+	case FREQ_ATTR_ID:
+		ret = sscanf(buf, "%hu", &fexcit);
+		if (ret != 1)
+			return -EINVAL;
+
+		ret = ad2s1210_reinit_excitation_frequency(dev, fexcit);
+		if (ret)
+			return ret;
+		break;
+
+	case HYSTERESIS_ATTR_ID:
+		ret = sscanf(buf, "%hhu", &hysteresis);
+		if (ret != 1)
+			return -EINVAL;
+
+		ret = ad2s1210_set_hysteresis(dev, hysteresis ? true : false);
+		if (ret)
+			return ret;
+
+		break;
+	default:
+		break;
+	}
+
+	return len;
+}
+
+/*!
+ * @brief       Read the debug register value
+ * @param       dev[in, out]- Pointer to IIO device instance
+ * @param       reg[in]- Register address to read from
+ * @param       readval[out]- Pointer to variable to read data into
+ * @return      0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad2s1210_debug_reg_read(void *dev, uint32_t reg,
+		uint32_t *val)
+{
+	return ad2s1210_reg_read(dev,  (uint8_t) reg, (uint8_t *)val);
+}
+
+/*!
+ * @brief       Write the debug register value
+ * @param       dev[in, out]- Pointer to IIO device instance
+ * @param       reg[in]- Register address to read from
+ * @param       writeval[out]- Pointer to variable to write data into
+ * @return      0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad2s1210_debug_reg_write(void *dev, uint32_t reg,
+		uint32_t val)
+{
+	return ad2s1210_reg_write(dev, (uint8_t)reg, (uint8_t)val);
+}
+
+/**
+ * @brief	Read buffer data corresponding to AD2S1210 IIO device
+ * @param	iio_dev_data[in] - Pointer to IIO device data structure
+ * @return	0 in case of success, negative error code otherwise
+ */
+static int32_t iio_ad2s1210_submit_buffer(struct iio_device_data *iio_dev_data)
+{
+#if (DATA_CAPTURE_MODE == BURST_DATA_CAPTURE)
+	enum ad2s1210_channel chn;
+	uint32_t sample_indx = 0;
+	uint32_t nb_of_samples;
+	uint16_t data;
+	int32_t ret;
+	uint8_t i;
+
+	nb_of_samples = iio_dev_data->buffer->size / BYTES_PER_SAMPLE;
+
+	if (!buf_size_updated) {
+		iio_dev_data->buffer->buf->size = iio_dev_data->buffer->size;
+		buf_size_updated = true;
+	}
+
+	while (sample_indx < nb_of_samples) {
+		/* Push into circular buffer */
+		for (i = 0; i < RESOLVER_CHANNELS; i++) {
+			if (ad2s1210_iio_channels[i].ch_type == IIO_ANGL)
+				chn = AD2S1210_POS;
+			else
+				chn = AD2S1210_VEL;
+
+			if (!(iio_dev_data->buffer->active_mask & (1 << i)))
```

Comment:
Would it be a good idea to do these operations in the  iio_ad2s1210_prepare_transfer? That way, we need not repeat these operations for every cycle. It also reduces the time the processor would take to read one sample

Reply:
I dont see a super optimization here?. ad2s1210_spi_single_conversion will only read one "channel" at the time . (with this hw we cannot read the 2 channels with the same spi operation). is the bit check that you find expensive? AFAICT  we will always have to do one check anyways to see which channel is/are enabled to push the right data into the buffer,